contracts/extensions/ERC721AQueries.sol

```diff
@@ -0,0 +1,158 @@
+    /**
+     * @dev Returns the `TokenOwnership` struct at `tokenId` without reverting.
+     *
+     * If the `tokenId` is out of bounds:
+     *   - `addr` = `address(0)`
+     *   - `startTimestamp` = `0`
+     *   - `burned` = `false`
+     *
+     * If the `tokenId` is burned:
+     *   - `addr` = `<Address of owner before token was burned>`
+     *   - `startTimestamp` = `<Timestamp when token was burned>`
+     *   - `burned = `true`
+     *
+     * Otherwise:
+     *   - `addr` = `<Address of owner>`
+     *   - `startTimestamp` = `<Timestamp of start of ownership>`
+     *   - `burned = `false`
+     */
+    function rawOwnershipOf(uint256 tokenId) public view returns (TokenOwnership memory) {
+        TokenOwnership memory ownership;
+        if (tokenId < _startTokenId() || tokenId >= _currentIndex) {
+            return ownership;
+        }
+        ownership = _ownerships[tokenId];
+        if (ownership.burned) {
+            return ownership;
+        }
+        return _ownershipOf(tokenId);
+    }
+
+    /**
+     * @dev Returns an array of `TokenOwnership` structs at `tokenIds` in order.
+     * See {ERC721AQueries-rawOwnershipOf}
+     */
+    function rawOwnershipsOf(uint256[] memory tokenIds) external view returns (TokenOwnership[] memory) {
+        unchecked {
+            uint256 tokenIdsLength = tokenIds.length;
+            TokenOwnership[] memory ownerships = new TokenOwnership[](tokenIdsLength);
+            for (uint256 i; i != tokenIdsLength; ++i) {
+                ownerships[i] = rawOwnershipOf(tokenIds[i]);
+            }
+            return ownerships;
+        }
+    }
+
+    /**
+     * @dev Returns an array of token IDs owned by `owner`,
+     * in the range [`start`, `stop`)
+     * (i.e. `start <= tokenId < stop`).
+     *
+     * This function allows for tokens to be queried if the collection
+     * grows too big for a single call of {ERC721AQueries-tokensOfOwner}.
+     */
+    function tokensOfOwnerIn(
+        address owner,
+        uint256 start,
+        uint256 stop
+    ) external view returns (uint256[] memory) {
+        unchecked {
+            uint256 tokenIdsIdx;
+            uint256 stopLimit = _currentIndex;
+            // Set `start = max(start, _startTokenId())`.
+            if (start < _startTokenId()) {
+                start = _startTokenId();
+            }
+            // Set `stop = min(stop, _currentIndex)`.
+            if (stop > stopLimit) {
+                stop = stopLimit;
+            }
+            uint256 tokenIdsMaxLength = balanceOf(owner);
+            // Set `tokenIdsMaxLength = min(balanceOf(owner), stop - start)`,
+            // to cater for cases where `balanceOf(owner)` is too big.
+            if (start < stop) {
+                uint256 windowLength = stop - start;
+                if (windowLength < tokenIdsMaxLength) {
+                    tokenIdsMaxLength = windowLength;
+                }
+            } else {
+                tokenIdsMaxLength = 0;
```

Comment:
I think it'd be better to check if start >= stop at the top of the function and throw an error if it's true. These are invalid parameters so it's better to error out than return 0 semantically

Reply:
Just note that even if we early revert at the top, the rest of the function still needs the same checks.

After `start` and `stop` are clamped, we will still need to check if `(start < stop)`, and then set `tokenIdsMaxLength = 0`, so that an empty array can be returned early.

Reply:
I agree with @Vectorized, Most of these features are for front-end queries anyways, it would be nice to be returned an empty array to handle instead of an error IMO

Reply:
It might be easier to handle on the frontend if you gracefully handle the error, but it doesn't mean it's the right way to do this. If you look at most of the functions in OZ's contracts, they all error out whenever you pass in invalid input parameters.

@Vectorized why do you need to keep the rest of the function the same? You will definitely have to change some tests if you make this change, I wouldn't expect the existing tests to pass.

Reply:
If I remove the checks and set to zero after the clamping, the function can go into a near infinite loop.

Reply:
What's the exact scenario you're thinking of? I don't see why we still need the if-statement on line 88. Line 93 can never evaluate to true

Reply:
Suppose a collection with `_currentIndex` = 8, and we pass in `start` = 10 and `stop` = 11. After clamping, `start` will remain 10, and `stop` will be 8.

Reply:
I don’t want to add another custom error for the out of bounds scenario.

Reply:
On custom errors, I am on the side that they are a poorly implemented feature of solidity and should be avoided as much as possible. 

If this PR is too troublesome to satisfy all parties, I suggest reverting the LowCap extension and leave it to end users to do their own implementation.

Reply:
What about doing `if (start >= stop || start >= stopLimit || stop <= _startTokenId) revert InvalidQueryRange();` after line 80? Basically, do the input validation after you know the start index and stop limit.

Reply:
That will cause it to revert for empty collections, which is not desired.